compat/win32/fscache.c

```diff
@@ -92,13 +106,13 @@ static void fsentry_init(struct fsentry *fse, struct fsentry *list,
 }
 
 /*
- * Allocate an fsentry structure on the heap.
+ * Allocate an fsentry structure.
  */
-static struct fsentry *fsentry_alloc(struct fsentry *list, const char *name,
+static struct fsentry *fsentry_alloc(struct fscache *cache, struct fsentry *list, const char *name,
 		size_t len)
 {
 	/* overallocate fsentry and copy the name to the end */
-	struct fsentry *fse = xmalloc(sizeof(struct fsentry) + len + 1);
+	struct fsentry *fse = mem_pool_alloc(cache->mem_pool, sizeof(struct fsentry) + len + 1);
```

Comment:
This causes that segfault in my hands. I see that the `mem_pool` is discarded before hitting this line:

```
Thread 1 hit Breakpoint 1, fscache_enable (initial_size=1)
    at compat/win32/fscache.c:369
369             int result = 0;
(gdb) c
Continuing.

Thread 1 hit Breakpoint 3, mem_pool_discard (mem_pool=0x351b1a8,
    invalidate_memory=0) at mem-pool.c:60
60              trace_printf_key(&trace_mem_pool, "mem_pool (%p): discard (%"PRIuMAX") unused\n",
(gdb) bt
#0  mem_pool_discard (mem_pool=0x351b1a8, invalidate_memory=0)
    at mem-pool.c:60
#1  0x0000000000633a4d in fscache_clear (cache=0x351b268)
    at compat/win32/fscache.c:270
#2  0x00000000006340b8 in fscache_flush () at compat/win32/fscache.c:463
#3  0x00000000005227e2 in write_entry (ce=0x351b060,
    path=0x351b508 "test.txt", state=0x12ff330, to_tempfile=0) at entry.c:371
#4  0x0000000000522cd3 in checkout_entry (ce=0x351b060, state=0x12ff330,
    topath=0x0) at entry.c:481
#5  0x0000000000418de9 in checkout_paths (opts=0x12ff450, revision=0x0)
    at builtin/checkout.c:368
#6  0x000000000041b734 in cmd_checkout (argc=1, argv=0x1bb13f8, prefix=0x0)
    at builtin/checkout.c:1277
#7  0x0000000000402cf5 in run_builtin (p=0x667190 <commands+336>, argc=2,
    argv=0x1bb13f8) at git.c:418
#8  0x00000000004030b0 in handle_builtin (argc=2, argv=0x1bb13f8) at git.c:649
#9  0x00000000004032ae in run_argv (argcp=0x12ffdd0, argv=0x12ffd78)
    at git.c:701
#10 0x00000000004035ac in cmd_main (argc=2, argv=0x1bb13f8) at git.c:803
#11 0x00000000004b9b58 in mingw_main (argc=3, argv=0x1bb13f0)
    at common-main.c:45
#12 0x00000000004b9b0c in main (argc=3, argv=0x3510498) at common-main.c:26
```

and this line is hit in this stacktrace:

```
(gdb) c
Continuing.

Thread 1 received signal SIGSEGV, Segmentation fault.
0x000000000065a396 in mem_pool_alloc (mem_pool=0x351b1a8, len=112)
    at mem-pool.c:87
87                  mem_pool->mp_block->end - mem_pool->mp_block->next_free >= len)
(gdb) bt
#0  0x000000000065a396 in mem_pool_alloc (mem_pool=0x351b1a8, len=112)
    at mem-pool.c:87
#1  0x00000000006333ee in fsentry_alloc (cache=0x351b268, list=0x0,
    name=0x351b0bc "test.txt", len=0) at compat/win32/fscache.c:115
#2  0x00000000006338ff in fsentry_create_list (cache=0x351b268,
    dir=0x12fef70, dir_not_found=0x12feefc) at compat/win32/fscache.c:229
#3  0x0000000000633c24 in fscache_get (cache=0x351b268, key=0x12fefd8)
    at compat/win32/fscache.c:323
#4  0x000000000063421a in fscache_lstat (filename=0x351b0bc "test.txt",
    st=0x12ff130) at compat/win32/fscache.c:493
#5  0x0000000000522839 in write_entry (ce=0x351b060,
    path=0x351b508 "test.txt", state=0x12ff330, to_tempfile=0) at entry.c:376
#6  0x0000000000522cd3 in checkout_entry (ce=0x351b060, state=0x12ff330,
    topath=0x0) at entry.c:481
#7  0x0000000000418de9 in checkout_paths (opts=0x12ff450, revision=0x0)
    at builtin/checkout.c:368
#8  0x000000000041b734 in cmd_checkout (argc=1, argv=0x1bb13f8, prefix=0x0)
    at builtin/checkout.c:1277
#9  0x0000000000402cf5 in run_builtin (p=0x667190 <commands+336>, argc=2,
    argv=0x1bb13f8) at git.c:418
#10 0x00000000004030b0 in handle_builtin (argc=2, argv=0x1bb13f8) at git.c:649
#11 0x00000000004032ae in run_argv (argcp=0x12ffdd0, argv=0x12ffd78)
    at git.c:701
#12 0x00000000004035ac in cmd_main (argc=2, argv=0x1bb13f8) at git.c:803
#13 0x00000000004b9b58 in mingw_main (argc=3, argv=0x1bb13f0)
    at common-main.c:45
#14 0x00000000004b9b0c in main (argc=3, argv=0x3510498) at common-main.c:26
```

I will have to take a break now, but maybe this makes any sort of sense to you?

Reply:
Oh, I guess that it is a bug that `fscache_clear()` discards the `mem_pool`. It should not. It should keep the `mem_pool`, it should just release the allocated memory, is all.

Reply:
Well, that is embarrassing.  Thank you for not only finding my bug but pointing me to the fix.  I'll have an updated PR as soon as the test suite finishes running twice (with and without GIT_TEST_FSCACHE=true)...